### PR TITLE
Add call-home agent feature

### PR DESCRIPTION
**Pull request by ctienshi**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    </dependencies>
+    <properties>
+        <Agent-Class>org.wso2.callhome.Agent</Agent-Class>
+    </properties>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-assembly-plugin</artifactId>
+<!--                <version>${maven.assembly.plugin.version}</version>-->
```
Comment (jsdjayanga) [inline pom.xml]: Why this line is commented?

File: src/main/java/org/wso2/callhome/CallHome.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                if (retry > 0) {
+                    logger.fine("Retry " + retry + "/" + 3);
+                    Thread.sleep(10000);
+                }
+                connection = (HttpsURLConnection) url.openConnection();
+                connection.setRequestProperty("Content-Type", "application/json");
+                connection.setRequestProperty("Accept", "application/json");
+                connection.setRequestProperty("Authorization", "Bearer " + ACCESS_TOKEN);
+                connection.setRequestMethod("GET");
+                connection.setConnectTimeout(3000);
```
Comment (jsdjayanga) [inline src/main/java/org/wso2/callhome/CallHome.java]: Connection time out of 3 seconds would not be enough. Shall we increase this to at least 10 seconds? (The delays may be due to Network or Backend service time)

File: src/main/java/org/wso2/callhome/utils/ExtractedInfo.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ */
+package org.wso2.callhome.utils;
+
+/**
+ * The ExtractedInfo class contains the information required for the call-home feature.
+ *
+ * @since 1.0.0
+ */
+public class ExtractedInfo {
+
```
Comment (chirangaalwis) [inline src/main/java/org/wso2/callhome/utils/ExtractedInfo.java]: @ctienshi can you please remove this extra, new line?

File: src/main/java/org/wso2/callhome/exception/CallHomeException.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Throws when an error is occurred in the CallHome feature.
+ */
+public class CallHomeException extends Exception {
+    public CallHomeException(String message) {
+        super(message);
+    }
+
+//    public CallHomeException(Throwable cause) {
```
Comment (chirangaalwis) [inline src/main/java/org/wso2/callhome/exception/CallHomeException.java]: @ctienshi @jsdjayanga do we need to maintain such an unused constructor (commented out), if not used anywhere?

File: src/main/java/org/wso2/callhome/exception/CallHomeException.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.callhome.exception;
+
+/**
+ * Throws when an error is occurred in the CallHome feature.
```
Comment (chirangaalwis) [inline src/main/java/org/wso2/callhome/exception/CallHomeException.java]: @ctienshi missing `@since` annotation in Java Docs.

**Comment by chirangaalwis** [comment]
@ctienshi since, this is a new repository and a new feature (if I understand correctly), shall we please create the relevant Git issues and release milestones, where necessary for the tasks you are fulfilling?

File: pom.xml (hunk 1 above)
Comment (ctienshi) [reply pom.xml]: I have deleted it and updated.

File: src/main/java/org/wso2/callhome/CallHome.java (hunk 2 above)
Comment (ctienshi) [reply src/main/java/org/wso2/callhome/CallHome.java]: Fixed and changed it to 10 seconds.

File: src/main/java/org/wso2/callhome/utils/ExtractedInfo.java (hunk 3 above)
Comment (ctienshi) [reply src/main/java/org/wso2/callhome/utils/ExtractedInfo.java]: Done

File: src/main/java/org/wso2/callhome/exception/CallHomeException.java (hunk 4 above)
Comment (ctienshi) [reply src/main/java/org/wso2/callhome/exception/CallHomeException.java]: Deleted it and updated.

File: src/main/java/org/wso2/callhome/exception/CallHomeException.java (hunk 5 above)
Comment (ctienshi) [reply src/main/java/org/wso2/callhome/exception/CallHomeException.java]: Added the annotation.

File: src/main/java/org/wso2/callhome/Agent.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+package org.wso2.callhome;
+
+/**
+ * This agent connects to the WSO2 Update servers to check whether there are new
+ * updates available for the running product and notify users accordingly.
+ *
+ * @since 1.0.0
+ */
+public class Agent {
+
```
Comment (chirangaalwis) [inline src/main/java/org/wso2/callhome/Agent.java]: @ctienshi please remove this additional new line.

Comment (ctienshi) [reply src/main/java/org/wso2/callhome/Agent.java]: Fixed.

File: src/main/java/org/wso2/callhome/CallHome.java [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+            logger.warning("Failed to get the number of updates available.");
+            logger.fine(e.toString());
+        }
+    }
+
+    /**
+     * This method returns the product home path.
+     *
+     * @return String The product home path
+     */
```
Comment (kasunsiyambalapitiya) [inline src/main/java/org/wso2/callhome/CallHome.java]: Shall we add the exception thrown to the doc comment?

File: src/main/java/org/wso2/callhome/CallHome.java [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+        return lastUpdateLevel;
+    }
+
+    /**
+     * This method constructs the URL from the given {@link ExtractedInfo} object.
+     *
+     * @param extractedInfo ExtractedInfo object
+     * @return Endpoint URL
+     */
+    private URL constructCallHomeURL(ExtractedInfo extractedInfo) throws CallHomeException {
```
Comment (kasunsiyambalapitiya) [inline src/main/java/org/wso2/callhome/CallHome.java]: Same as `getProductHome` method, shall we add the exception thrown to the doc comment?

File: src/main/java/org/wso2/callhome/CallHome.java (hunk 7 above)
Comment (ctienshi) [reply src/main/java/org/wso2/callhome/CallHome.java]: Fixed.

File: src/main/java/org/wso2/callhome/CallHome.java (hunk 8 above)
Comment (ctienshi) [reply src/main/java/org/wso2/callhome/CallHome.java]: Fixed.

**Pull request merged**